src/XR/features/WebXREyeTracking.ts

```diff
@@ -0,0 +1,151 @@
+    constructor(_xrSessionManager: WebXRSessionManager) {
+        super(_xrSessionManager);
+        this.xrNativeFeatureName = "eye-tracking";
+        if (this._xrSessionManager.session) {
+            this._init();
+        } else {
+            this._xrSessionManager.onXRSessionInit.addOnce(() => {
+                this._init();
+            });
+        }
+    }
+
+    /**
+     * Dispose this feature and all of the resources attached.
+     */
+    public dispose(): void {
+        super.dispose();
+
+        Object.keys(this._eventListeners).forEach((eventName: string) => {
+            const func = this._eventListeners && this._eventListeners[eventName as XREventType];
+            if (func) {
+                this._xrSessionManager.session.removeEventListener(eventName as XREventType, func);
+            }
+        });
+
+        this.onEyeTrackingStartedObservable.clear();
+        this.onEyeTrackingEndedObservable.clear();
+        this.onEyeTrackingFrameUpdateObservable.clear();
+    }
+
+    /**
+     * Returns whether the gaze data is valid or not
+     * @returns true if the data is valid
+     */
+    public isEyeGazeValid(): boolean {
+        return !!this._gazeRay;
+    }
+
+    /**
+     * Get a reference to the gaze ray. This data is valid while eye tracking persists, and will be set to null when gaze data is no longer available
+     * @returns a reference to the gaze ray if it exists and is valid, returns null otherwise.
+     */
+    public getEyeGaze(): Nullable<Ray> {
+        return this._gazeRay;
+    }
+
+    protected _onXRFrame(frame: XRFrame) {
+        if (!this.attached || !frame) {
+            return;
+        }
+
+        if (this._latestEyeSpace && this._gazeRay) {
+            let pose = frame.getPose(this._latestEyeSpace, this._xrSessionManager.referenceSpace);
+            if (pose) {
+                this._gazeRay.origin.set(pose.transform.position.x, pose.transform.position.y, pose.transform.position.z);
+                const quat = pose.transform.orientation;
+                this._tmpQuat.set(quat.x, quat.y, quat.z, quat.w);
+
+                if (!this._xrSessionManager.scene.useRightHandedSystem) {
+                    this._gazeRay.origin.z *= -1;
+                    this._tmpQuat.z *= -1;
+                    this._tmpQuat.w *= -1;
+                }
+
+                Vector3.Forward().rotateByQuaternionToRef(this._tmpQuat, this._gazeRay.direction);
+                this.onEyeTrackingFrameUpdateObservable.notifyObservers(this._gazeRay);
+            }
+        }
+    }
+
+    private _init() {
+        // Only supported by BabylonNative
+        if (!!this._xrSessionManager.isNative) {
+
+            const eyeTrackingStartListener = (event: XREyeTrackingSourceEvent) => {
+                this._latestEyeSpace = event.gazeSpace;
+                this._gazeRay = new Ray(Vector3.Zero(), Vector3.Forward());
+                this.onEyeTrackingStartedObservable.notifyObservers(this._gazeRay);

```

Comment:
should this actually be straight forward relative to the xr camera?

Reply:
Since it's a placeholder until we get a frame update I wasn't too concerned about what direction it was in. From what I can see, I'd need a frame object in order to get an accurate HMD direction from the XRSessionManager's viewerReferenceSpace, at which point we get tracked eye positions.